### PR TITLE
feat(index): merge `ruleContexts` search parameter

### PR DESCRIPTION
**Pull request by francoischalifour**
This adds support for the [`ruleContexts`](https://www.algolia.com/doc/api-reference/api-parameters/ruleContexts/) search parameter in the merge function.

### CODE REVIEW

File: src/lib/utils/__tests__/mergeSearchParameters-test.ts [hunk 1]
```diff
@@ -367,4 +367,31 @@ describe('mergeSearchParameters', () => {
       })
     );
   });
+
+  it('merges and dedupes `ruleContexts` parameters', () => {
+    const actual = merge(
+      algoliasearchHelper.SearchParameters.make({
+        ruleContexts: ['ais-genre-comedy'],
```
Comment (francoischalifour) [inline src/lib/utils/__tests__/mergeSearchParameters-test.ts]: The `make` function is not yet properly typed in the JS helper at this point and is not aware of the `ruleContexts` search parameter. Rebasing should work.

Comment (Haroenv) [reply src/lib/utils/__tests__/mergeSearchParameters-test.ts]: see https://github.com/algolia/instantsearch.js/pull/3940 which will have a fix for that

File: src/lib/utils/mergeSearchParameters.ts [hunk 2]
```diff
@@ -103,6 +105,23 @@ const mergeHierarchicalFacetsRefinements: Merger = (left, right) =>
...
 
+const mergeRuleContexts: Merger = (left, right) => {
+  const ruleContexts: string[] = uniq(
+    ([] as any)
+      .concat(left.ruleContexts)
+      .concat(right.ruleContexts)
+      .filter(Boolean)
+  );
+
+  if (ruleContexts.length > 0) {
```
Comment (francoischalifour) [inline src/lib/utils/mergeSearchParameters.ts]: Note that I add the `ruleContexts` key in the search parameters only when not empty. Do we want to always add it?

File: src/lib/utils/mergeSearchParameters.ts [hunk 3]
```diff
@@ -103,6 +105,23 @@ const mergeHierarchicalFacetsRefinements: Merger = (left, right) =>
     },
   });
 
+const mergeRuleContexts: Merger = (left, right) => {
+  const ruleContexts: string[] = uniq(
+    ([] as any)
+      .concat(left.ruleContexts)
+      .concat(right.ruleContexts)
+      .filter(Boolean)
```
Comment (Haroenv) [inline src/lib/utils/mergeSearchParameters.ts]: why do you need to filter empty values out here?

Comment (francoischalifour) [reply src/lib/utils/mergeSearchParameters.ts]: `right.ruleContexts` might be `undefined`.

File: src/lib/utils/mergeSearchParameters.ts (hunk 2 above)
Comment (samouss) [reply src/lib/utils/mergeSearchParameters.ts]: I would say omit when it doesn't make sense.

**Pull request merged**